zk/stages/stage_witness.go

```diff
@@ -0,0 +1,562 @@
+		storeNblocks:       storeNblocks,
+		witnessFull:        witnessFull,
+	}
+}
+
+func SpawnWitnessStage(
+	s *stagedsync.StageState,
+	u stagedsync.Unwinder,
+	tx kv.RwTx,
+	ctx context.Context,
+	cfg WitnessCfg,
+	initialCycle bool,
+	quiet bool,
+) error {
+	logPrefix := s.LogPrefix()
+	if !cfg.enableWitnessStage {
+		log.Info(fmt.Sprintf("[%s] Skipping witness stage", logPrefix))
+		return nil
+	}
+
+	log.Info(fmt.Sprintf("[%s] Starting witness stage", logPrefix))
+	defer log.Info(fmt.Sprintf("[%s] Finished witness stage", logPrefix))
+
+	if sequencer.IsSequencer() {
+		log.Info(fmt.Sprintf("[%s] skipping -- sequencer", logPrefix))
+		return nil
+	}
+
+	if tx == nil {
+		log.Debug(fmt.Sprintf("[%s] witness: no tx provided, creating a new one", logPrefix))
+		var err error
+		tx, err = cfg.db.BeginRw(ctx)
+		if err != nil {
+			return fmt.Errorf("failed to open tx, %w", err)
+		}
+		defer tx.Rollback()
+	}
+
+	// batch until which witness has been generated and saved to db
+	witnessProgress, err := stages.GetStageProgress(tx, stages.Witness)
+	if err != nil {
+		return fmt.Errorf("save stage progress error: %v", err)
+	}
+
+	// batch which has been processed by stage_batches
+	latestBlock, err := stages.GetStageProgress(tx, stages.Execution)
+	if err != nil {
+		return fmt.Errorf("could not retrieve batch no progress")
+	}
+
+	hermezDbReader := hermez_db.NewHermezDbReader(tx)
+
+	batchesProgress, err := hermezDbReader.GetBatchNoByL2Block(latestBlock)
+	if err != nil {
+		return fmt.Errorf("could not retrieve batch for latest block")
+	}
+
+	if witnessProgress >= batchesProgress {
+		log.Info("All batch witness already in db, skipping")
+		return nil
+	}
+
+	// witness generation should happen till batch of latest block -1
+	// as there might be more blocks incoming for this batch
+	witnessToBatch := batchesProgress - 1
+
+	highestBlockInLatestBatch, err := hermezDbReader.GetHighestBlockInBatch(witnessToBatch)
+	if err != nil {
+		return err
+	}
+
+	blockNblocksBehind := highestBlockInLatestBatch - cfg.storeNblocks + 1
+	witnessFromBatch, err := hermezDbReader.GetBatchNoByL2Block(blockNblocksBehind)
+	if err != nil {
+		return err
+	}
+
+	wg := NewWitnessGenerator(cfg.tmpDir, cfg.historyV3, cfg.agg, cfg.blockReader, cfg.chainConfig, cfg.engine)
+
+	for b := witnessFromBatch; b <= witnessToBatch; b++ {
+		highestBlockInBatch, err := hermezDbReader.GetHighestBlockInBatch(b)
+		if err != nil {
+			return err
+		}
+
+		if highestBlockInBatch == 0 {
+			continue
+		}
+
+		nextStageProgress, err := stages.GetStageProgress(tx, stages.HashState)
+		if err != nil {
+			return err
+		}
+		if nextStageProgress < highestBlockInBatch {
+			// Skip this stage
+			return nil
+		}
+
+		lowestBlockInBatch, err := hermezDbReader.GetLowestBlockInBatch(b)
+		if err != nil {
+			return err
+		}
+
+		// Skip witness generation for this batch as it is bound to fail
+		if highestBlockInBatch-lowestBlockInBatch > maxGetProofRewindBlockCount {
```

Comment:
@revitteth right I'm skipping the witness stage if it has moved past 1000 blocks

Reply:
I think instead of `highestBlockInBatch` I should have `latestBlock` there, that will fix the mentioned issue. Will try on my end and sync up cardona till tip to test it.